src/graphnet/models/components/layers.py

```diff
@@ -593,3 +602,393 @@ def forward(
+            )
+
+        if self.rezero:  # Learnable scaling parameters
+            self.alpha1_x = nn.Parameter(torch.zeros(1, 1))
+            self.alpha2_x = nn.Parameter(torch.zeros(1, 1))
+            self.alpha1_e = nn.Parameter(torch.zeros(1, 1))
+
+        self.dropout1 = nn.Dropout(dropout)  # Post-attention dropout on x
+        self.dropout2 = nn.Dropout(dropout)  # Post-attention dropout on e
+        self.dropout3 = nn.Dropout(dropout)  # Post-FFN dropout on x
+
+    def forward(self, data: Data) -> Data:
+        """Forward pass."""
+        x = data.x
+        num_nodes = data.num_nodes
+        log_deg = get_log_deg(data)
+
+        x_attn_residual = x  # for first residual connection
+        e_values_in = data.get("edge_attr", None)
+        e = None
+
+        # Attention outputs
+        x_attn_out, e_attn_out = self.attention(data)
+
+        x = x_attn_out.view(num_nodes, -1)
+        x = self.dropout1(x)
+
+        # Apply degree scaler if enabled
+        if self.deg_scaler:
+            x = torch.stack([x, x * log_deg], dim=-1)
+            x = (x * self.deg_coef).sum(dim=-1)
+
+        x = self.fc1_x(x)
+        if e_attn_out is not None:
+            e = e_attn_out.flatten(1)
+            e = self.dropout2(e)
+            e = self.fc1_e(e)
+
+        if self.residual:
+            if self.rezero:
+                x = x * self.alpha1_x
+            x = x_attn_residual + x
+
+            if e is not None:
+                if self.rezero:
+                    e = e * self.alpha1_e
+                e = e + e_values_in
+
+        x = self.norm1_x(x)
+        if e is not None:
+            e = self.norm1_e(e)
+
+        # FFN for x
+        x_ffn_residual = x  # Residual over the FFN
+        x = self.FFN_x_layer1(x)
+        x = self.activation(x)
+        x = self.dropout3(x)
+        x = self.FFN_x_layer2(x)
+
+        if self.residual:
+            if self.rezero:
+                x = x * self.alpha2_x
+            x = x_ffn_residual + x  # residual connection
+
+        x = self.norm2_x(x)
+
+        data.x = x
+        if self.update_edges:
+            data.edge_attr = e
+        else:
+            data.edge_attr = e_values_in
+
+        return data
+
+
+# TODO: This is a prediction head... we probably want only the graph stuff here
+# and let the Tasks handle the last layer. -PW
+class SANGraphHead(LightningModule):
+    """SAN prediction head for graph prediction tasks.
+
+    Original code:
+    https://github.com/LiamMa/GRIT/blob/main/grit/head/san_graph.py
+    """
+
+    def __init__(
+        self,
+        dim_in: int,
+        dim_out: int = 1,
+        L: int = 2,
+        activation: nn.Module = nn.ReLU,
+        pooling: str = "mean",
+    ):
+        """Construct `SANGraphHead`.
+
+        Args:
+            dim_in: Input dimension.
+            dim_out: Output dimension.
+            L: Number of hidden layers.
+            activation: Activation function.
```

Comment:
Could mention here that the activation function is assumed to be un-instantiated. E.g.
```
"""
activation:  Reference to uninstantiated activation function. E.g. `torch.nn.LeakyReLU`
"""
```